Detect Windows & Mac OS version.

carrgo opened the pull request.

zshenker commented:
@fnando Any chance of getting this merged in? Would be very useful.

@carrgo Any thoughts as to being able to access the major/minor/patch parts of the version separately for Mac? 
i.e. I have a case where I want to make sure that the OS X version is >= 10.10.3

fnando commented on lib/browser/platform/windows.rb:
```diff
@@ -2,8 +2,24 @@
       def version
-        "0"
+        nt_version = ua[/Windows(?:\sNT)?\s*([0-9_\.]+)?/, 1]
+        VERSION_MAP[nt_version] || "0"
```
I don't think `Browser::Platform::Windows#version` should return a readable version; it must a numeric/float-based version (pretty much the key value).

Otherwise one can't do version comparison, like in `[">=10"]`, similar to other version comparison methods from this library.

carrgo replied:
In that case, would you use the key version extracted e.g. 6.3 for Windows 8.1 or replace that with 8.1? And what about the other version that aren't really numeric e.g. Vista/XP/2000?

fnando commented on lib/browser/platform/mac.rb:
```diff
@@ -3,7 +3,7 @@ module Browser
     class Mac < Base
       def version
-        "0"
+        ua[/Mac OS X\s*([0-9_\.]+)?/, 1].gsub('_', '.')
```
Use double quotes.

fnando commented on lib/browser/platform/windows.rb:
```diff
@@ -2,8 +2,24 @@
+        "5.01" => "2000, Service Pack 1 (SP1)",
+        "5.0"  => "2000",
+        "4.0"  => "NT 4.0",
+        "98"   => "98",
```
Numeric version for Win98: https://en.wikipedia.org/wiki/Windows_98#Development

fnando commented on lib/browser/platform/windows.rb:
```diff
@@ -2,8 +2,24 @@
+        "5.0"  => "2000",
+        "4.0"  => "NT 4.0",
+        "98"   => "98",
+        "95"   => "95",
```
Numeric version for Win95: https://en.wikipedia.org/wiki/Windows_95#Editions

fnando commented:
In general terms, I'm okay with this. The main concern here is the Window versioning, which should be numeric/float-based, similar to everything else.

> i.e. I have a case where I want to make sure that the OS X version is >= 10.10.3

I think a companion method should be introduced, so that you could do something like `browser.platform.version?([">=10.10.3"])`.

carrgo commented:
@fnando As per your comment, I've updated to use double quotes. Also I've updated the Windows version to return the NT version. Unfortunately this meant removing support for 95/98 (but I hardly think anyone is still using them anyway).

I thought about updating the browser.plaform.windows*? methods to use the version but it got a bit messy and ended up performing a regex twice which I didn't think was very efficient. Up to you on how you want to proceed with this.

fnando commented:
@carrgo You can find Win95/98 versions here:

- https://en.wikipedia.org/wiki/Windows_95#Editions
- https://en.wikipedia.org/wiki/Windows_98#Development

I don't if it's really required though.

carrgo commented:
Yeah you can get the versions for Windows 95/98, however, 95/98 aren't part of the Windows NT stream and so those versions aren't exactly comparable to later releases of Windows. For example, both Windows 95 and Windows NT 4.0 are v4.0. Windows 98 and NT share some similarities but are still different product streams.

fnando commented:
Let's ignore 95/98 for now.
